Applications/Spire/Source/Ui/DropDownMenu2.cpp

```diff
@@ -0,0 +1,241 @@
+#include "Spire/Ui/DropDownMenu2.hpp"
+#include <QFocusEvent>
+#include <QKeyEvent>
+#include <QScrollBar>
+#include <QVBoxLayout>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ScrollArea.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  const auto BORDER_PADDING = 2;
+
+  void adjust_current_index(DropDownMenu2& menu, int amount) {
+    if(menu.count() == 0 || amount == 0) {
+      return;
+    }
+    auto last_index = menu.count() - 1;
+    auto index = [&] {
+      auto index = menu.get_current();
+      if(!index) {
+        if(amount > 0) {
+          return 0;
+        } else {
+          return last_index;
+        }
+      }
+      *index += amount;
+      if(index > last_index) {
+        return 0;
+      } else if(index < 0) {
+        return last_index;
+      }
+      return *index;
+    }();
+    menu.set_current(index);
+  }
+}
+
+DropDownMenu2::DropDownMenu2(std::vector<DropDownItem*> items,
+    QWidget* parent)
+    : DropDownWindow(true, parent),
+      m_selected_index(0),
+      m_max_displayed_items(5) {
+  m_scroll_area = new ScrollArea(this);
+  m_scroll_area->setFocusProxy(parent);
+  m_scroll_area->setWidgetResizable(true);
+  auto main_widget = new QWidget(this);
+  m_layout = new QVBoxLayout(main_widget);
+  m_layout->setContentsMargins({});
+  m_layout->setSpacing(0);
+  m_scroll_area->setWidget(main_widget);
+  initialize_widget(m_scroll_area);
+  for(auto& item : items) {
+    m_layout->addWidget(item);
+    m_item_hovered_connections.AddConnection(item->connect_highlighted_signal(
+      [=] (const auto& value) { m_hovered_signal(item->get_value()); }));
+    m_item_selected_connections.AddConnection(item->connect_selected_signal(
+      [=] (const auto& value) { on_item_selected(item->get_value(), item); }));
+  }
+  if(m_layout->count() > 0) {
+    update_height();
+    static_cast<DropDownItem*>(m_layout->itemAt(0)->widget())->set_highlight();
+    m_selected_index = 0;
+    m_current_index = 0;
+    m_selected_signal(get_value(*m_selected_index));
+    m_current_signal(get_value(*m_current_index));
+  }
+  parent->installEventFilter(this);
```

Comment:
This component should not be watching its parent or have anything to do with its parent.

Reply:
The DropDownMenu needs to listen to parent's events to get the up, down, and enter/return key events. The alternative would be to have the inputs define that behavior, but then it's repeated every time the DropDownMenu is used.

Reply:
The inputs define that behavior. As a general principle, avoid introducing the potential for "fights" between components where components have to install event listeners and override one another and basically constantly fight each other for control. To do this there is a clear hierarchy where a component only ever has control over its children and never crosses its API boundary.